File: apps/smahties/src/embedding.rs
use async_openai::{
    Client,
    config::OpenAIConfig,
    types::embeddings::{CreateEmbeddingRequestArgs, Embedding},
};
use std::ops::Range;

use crate::Result;

const DEFAULT_MAX_EMBEDDING_BATCH_INPUTS: usize = 128;
const DEFAULT_MAX_EMBEDDING_BATCH_BYTES: usize = 256 * 1024;

#[derive(Clone)]
pub struct OpenAiEmbedder {
    client: Client<OpenAIConfig>,
    model: String,
    batch_limits: EmbeddingBatchLimits,
}

#[derive(Clone, Copy, Debug)]
struct EmbeddingBatchLimits {
    max_inputs: usize,
    max_request_bytes: usize,
}

impl Default for EmbeddingBatchLimits {
    fn default() -> Self {
        Self {
            max_inputs: DEFAULT_MAX_EMBEDDING_BATCH_INPUTS,
            max_request_bytes: DEFAULT_MAX_EMBEDDING_BATCH_BYTES,
        }
    }
}

impl OpenAiEmbedder {
    pub fn new(base_url: &str, model: String) -> Self {
        let config = OpenAIConfig::new().with_api_base(base_url);
        Self {
            client: Client::with_config(config),
            model,
            batch_limits: EmbeddingBatchLimits::default(),
        }
    }

    pub fn model(&self) -> &str {
        &self.model
    }

    pub async fn embed_texts(&self, texts: &[String]) -> Result<Vec<Vec<f32>>> {
        if texts.is_empty() {
            return Ok(Vec::new());
        }

        let mut embeddings = Vec::with_capacity(texts.len());
        for range in embedding_batch_ranges(texts, self.batch_limits)? {
            let mut batch_embeddings = self.embed_text_batch(&texts[range]).await?;
            embeddings.append(&mut batch_embeddings);
        }

        Ok(embeddings)
    }

    async fn embed_text_batch(&self, texts: &[String]) -> Result<Vec<Vec<f32>>> {
        let request = CreateEmbeddingRequestArgs::default()
            .model(&self.model)
            .input(texts.to_vec())
            .build()?;
        let response = self.client.embeddings().create(request).await?;

        let embeddings = ordered_embeddings(response.data)?;
        if embeddings.len() != texts.len() {
            return Err(crate::SmahtiesError::InvalidRequest(format!(
                "embedding response count {} did not match batch input count {}",
                embeddings.len(),
                texts.len()
            )));
        }
        Ok(embeddings)
    }
}

fn embedding_batch_ranges(
    texts: &[String],
    limits: EmbeddingBatchLimits,
) -> Result<Vec<Range<usize>>> {
    if limits.max_inputs == 0 || limits.max_request_bytes == 0 {
        return Err(crate::SmahtiesError::InvalidRequest(
            "embedding batch limits must be greater than zero".into(),
        ));
    }

    let mut ranges = Vec::new();
    let mut start = 0;
    let mut request_bytes = 0;

    for (index, text) in texts.iter().enumerate() {
        let batch_inputs = index - start;
        let would_exceed_inputs = batch_inputs >= limits.max_inputs;
        let would_exceed_bytes =
            batch_inputs > 0 && request_bytes + text.len() > limits.max_request_bytes;

        if would_exceed_inputs || would_exceed_bytes {
            ranges.push(start..index);
            start = index;
            request_bytes = 0;
        }

        request_bytes += text.len();
    }

    if start < texts.len() {
        ranges.push(start..texts.len());
    }

    Ok(ranges)
}

fn ordered_embeddings(mut embeddings: Vec<Embedding>) -> Result<Vec<Vec<f32>>> {
    embeddings.sort_by_key(|embedding| embedding.index);
    for (expected, embedding) in embeddings.iter().enumerate() {
        if embedding.index as usize != expected {
            return Err(crate::SmahtiesError::InvalidRequest(format!(
                "embedding response index {} did not match expected index {expected}",
                embedding.index
            )));
        }
    }

    Ok(embeddings
        .into_iter()
        .map(|embedding| embedding.embedding)
        .collect())
}

#[cfg(test)]
mod tests {
    use async_openai::types::embeddings::Embedding;

    use super::{EmbeddingBatchLimits, embedding_batch_ranges, ordered_embeddings};

    #[test]
    fn embeddings_are_returned_in_request_order() {
        let embeddings = ordered_embeddings(vec![
            Embedding {
                index: 1,
                object: "embedding".into(),
                embedding: vec![1.0],
            },
            Embedding {
                index: 0,
                object: "embedding".into(),
                embedding: vec![0.0],
            },
        ])
        .unwrap();

        assert_eq!(embeddings, vec![vec![0.0], vec![1.0]]);
    }

    #[test]
    fn embeddings_reject_missing_response_indexes() {
        let error = ordered_embeddings(vec![Embedding {
            index: 1,
            object: "embedding".into(),
            embedding: vec![1.0],
        }])
        .unwrap_err();

        assert!(error.to_string().contains("expected index 0"));
    }

    #[test]
    fn embedding_batches_are_limited_by_input_count() {
        let texts = ["one", "two", "three", "four", "five"].map(String::from);
        let ranges = embedding_batch_ranges(
            &texts,
            EmbeddingBatchLimits {
                max_inputs: 2,
                max_request_bytes: usize::MAX,
            },
        )
        .unwrap();

        assert_eq!(ranges, vec![0..2, 2..4, 4..5]);
    }

    #[test]
    fn embedding_batches_are_limited_by_request_bytes() {
        let texts = ["aa", "bb", "cc"].map(String::from);
        let ranges = embedding_batch_ranges(
            &texts,
            EmbeddingBatchLimits {
                max_inputs: 10,
                max_request_bytes: 5,
            },
        )
        .unwrap();

        assert_eq!(ranges, vec![0..2, 2..3]);
    }

    #[test]
    fn oversized_single_inputs_are_sent_alone() {
        let texts = ["abcdef", "g"].map(String::from);
        let ranges = embedding_batch_ranges(
            &texts,
            EmbeddingBatchLimits {
                max_inputs: 10,
                max_request_bytes: 5,
            },
        )
        .unwrap();

        assert_eq!(ranges, vec![0..1, 1..2]);
    }

    #[test]
    fn embedding_batches_reject_zero_limits() {
        let texts = ["one"].map(String::from);
        let error = embedding_batch_ranges(
            &texts,
            EmbeddingBatchLimits {
                max_inputs: 0,
                max_request_bytes: 1,
            },
        )
        .unwrap_err();

        assert!(error.to_string().contains("greater than zero"));
    }
}

File: src/smahties/embedding.py
from __future__ import annotations

import os
import asyncio
from dataclasses import dataclass

from openai import AsyncOpenAI

DEFAULT_MAX_EMBEDDING_BATCH_INPUTS = 128
DEFAULT_MAX_EMBEDDING_BATCH_BYTES = 256 * 1024
DEFAULT_MAX_EMBEDDING_CONCURRENCY = 2


@dataclass(frozen=True)
class EmbeddingBatchLimits:
    """Maximum OpenAI embedding request size limits."""

    max_inputs: int = DEFAULT_MAX_EMBEDDING_BATCH_INPUTS
    max_request_bytes: int = DEFAULT_MAX_EMBEDDING_BATCH_BYTES
    max_concurrent_requests: int = DEFAULT_MAX_EMBEDDING_CONCURRENCY


class OpenAiEmbedder:
    """OpenAI-compatible embedding client with deterministic batching."""

    def __init__(
        self, base_url: str, model: str, limits: EmbeddingBatchLimits | None = None
    ) -> None:
        self.model = model
        self.client = AsyncOpenAI(
            base_url=base_url,
            api_key=os.environ.get("OPENAI_API_KEY", "not-needed"),
        )
        self.limits = limits or EmbeddingBatchLimits()

    async def embed_texts(self, texts: list[str]) -> list[list[float]]:
        """Embed texts in request-sized batches and preserve input order."""

        ranges = embedding_batch_ranges(texts, self.limits)
        if not ranges:
            return []
        if self.limits.max_concurrent_requests == 1 or len(ranges) == 1:
            embeddings: list[list[float]] = []
            for start, end in ranges:
                embeddings.extend(await self._embed_range(texts, start, end))
            return embeddings

        results: list[list[list[float]] | None] = [None] * len(ranges)
        next_range = 0

        async def worker() -> None:
            nonlocal next_range
            while next_range < len(ranges):
                range_index = next_range
                next_range += 1
                start, end = ranges[range_index]
                results[range_index] = await self._embed_range(texts, start, end)

        worker_count = min(self.limits.max_concurrent_requests, len(ranges))
        await asyncio.gather(*(worker() for _ in range(worker_count)))

        embeddings: list[list[float]] = []
        for result in results:
            if result is None:
                raise ValueError("embedding worker did not produce a result")
            embeddings.extend(result)
        return embeddings

    async def _embed_range(self, texts: list[str], start: int, end: int) -> list[list[float]]:
        """Embed one pre-sized input range and validate response ordering."""

        response = await self.client.embeddings.create(model=self.model, input=texts[start:end])
        batch = sorted(response.data, key=lambda item: item.index)
        if len(batch) != end - start:
            raise ValueError(
                f"embedding response count {len(batch)} did not match batch input count {end - start}"
            )
        embeddings: list[list[float]] = []
        for expected, item in enumerate(batch):
            if item.index != expected:
                raise ValueError(
                    f"embedding response index {item.index} did not match expected index {expected}"
                )
            embeddings.append(list(item.embedding))
        return embeddings


def embedding_batch_ranges(
    texts: list[str],
    limits: EmbeddingBatchLimits,
) -> list[tuple[int, int]]:
    """Return half-open input ranges that fit embedding request limits."""

    if (
        limits.max_inputs <= 0
        or limits.max_request_bytes <= 0
        or limits.max_concurrent_requests <= 0
    ):
        raise ValueError("embedding batch limits must be greater than zero")
    ranges: list[tuple[int, int]] = []
    start = 0
    request_bytes = 0
    for index, text in enumerate(texts):
        batch_inputs = index - start
        would_exceed_inputs = batch_inputs >= limits.max_inputs
        would_exceed_bytes = (
            batch_inputs > 0 and request_bytes + len(text) > limits.max_request_bytes
        )
        if would_exceed_inputs or would_exceed_bytes:
            ranges.append((start, index))
            start = index
            request_bytes = 0
        request_bytes += len(text)
    if start < len(texts):
        ranges.append((start, len(texts)))
    return ranges
